### PR TITLE
[API] Access Token endpoints

### PR DESCRIPTION
**Pull request by Martouta**
Fixes [THREESCALE-1766 - Improve the API endpoints of access token](https://issues.jboss.org/browse/THREESCALE-1766)

The system/ruby team had a meeting about this and I will explain here what we decided. We also informed product about this.

### What we have right now (before this PR)
We have only 1 API endpoint for access tokens but we have much more in the UI. The API endpoint that we have right now was created years ago before we were open source in https://github.com/3scale/system/pull/8441 but it was not documented and officially not supported and we documented it and made it public recently in https://github.com/3scale/porta/pull/430

This endpoint right now is under the URL `provider-url/admin/api/users/:user_id/access_tokens`. An officially documented that customers can authenticate through access token and we will allow to create only if the user_id belongs to the same user. Which is true but we also have (not documented and officially unsupported) the possibility to authenticate through provider_key and then we will allow to create for any user of that same account.

This endpoint returns the value that it just created, and the creation is the only moment when we return the value.

We also need to keep in mind that apicast may need to authenticate through provider_key 😄 

### What we we aim to have (with this PR)
We want it to be similar to what we have in the UI. So what we will do the next endpoints.

#### Index/Search Access Token API endpoint
Under the URL `provider-url/admin/api/personal/access_tokens`. It will authenticate through the access_token and will return all the access tokens of the same user of the access_token if no extra parameter is sent. We will also allow the user to send a `name` string query param (like `provider-url/admin/api/access_tokens?name=example`) and we will return all the access tokens of that current user which name contains that string of the param.

For each access token we will return the following attributes.
  - scopes
  - permission
  - name

We will **not** return their values for security reasons 😄 

##### New endpoint
image
Example of request without search: `curl -v  -X GET "http://provider-admin.example.com.lvh.me:3000/admin/api/personal/access_tokens.json?access_token=13d435a15183e848161ec41a3f5a41b3fee6f2f5bc41ddd0308884a7c42a98cd"`
Example of response:
```json
{
 "access_tokens": [
  {
   "access_token": {
    "id": 3,
    "name": "account",
    "scopes": [
     "account_management"
    ],
    "permission": "rw"
   }
  },
  {
   "access_token": {
    "id": 4,
    "name": "example",
    "scopes": [
     "finance",
     "stats"
    ],
    "permission": "ro"
   }
  },
  {
   "access_token": {
    "id": 5,
    "name": "new example and different",
    "scopes": [
     "account_management",
     "stats"
    ],
    "permission": "rw"
   }
  }
 ]
}
```
Example of request with search: `curl -v  -X GET "http://provider-admin.example.com.lvh.me:3000/admin/api/personal/access_tokens.json?name=xamp&access_token=13d435a15183e848161ec41a3f5a41b3fee6f2f5bc41ddd0308884a7c42a98cd"`

#### Show Access Token API endpoint
Under the URL `provider-url/admin/api/personal/access_tokens/ID` or `provider-url/admin/api/personal/access_tokens/VALUE`
It will authenticate through the access_token and only that same user can fetch an access token of its own. It can be fetched by `ID` or by its `value`.
We will return for the access token the same parameters as for the index (the previous endpoint).

##### New endpoint
image
Example of request: `curl -v  -X GET "http://provider-admin.example.com.lvh.me:3000/admin/api/personal/access_tokens/4.json?access_token=13d435a15183e848161ec41a3f5a41b3fee6f2f5bc41ddd0308884a7c42a98cd"`
Example of response:
```json
{
 "access_token": {
  "id": 4,
  "name": "example",
  "scopes": [
   "finance",
   "stats"
  ],
  "permission": "ro"
 }
}
```

#### Delete Access Token API endpoint
Under the URL `provider-url/admin/api/personal/access_tokens/ID` or `provider-url/admin/api/personal/access_tokens/VALUE`
It will authenticate through the access_token and only that same user can delete an access token of its own. It can be fetched by `ID` or by its `value`.

##### New endpoint
image
Example of request: `curl -v  -X DELETE "http://provider-admin.example.com.lvh.me:3000/admin/api/personal/access_tokens/5.json" -d 'access_token=13d435a15183e848161ec41a3f5a41b3fee6f2f5bc41ddd0308884a7c42a98cd'`

#### Create Access Token API endpoint
We will have a new endpoint that does exactly the same as we have, but with a different URL, under `provider-url/admin/api/personal/access_tokens`, and this new endpoint will allow to authenticate only through access_token and not through provider_key.
The endpoint that we had until now will keep working but will be deprecated.

We will also improve the documentation for both endpoints to have the following description:
> Creates an access token. Make sure to copy your new personal access token now. You will not be able to see it again as it is not stored for security reasons.

##### New endpoint
image
Example of request: `curl -v  -X POST "http://provider-admin.example.com.lvh.me:3000/admin/api/personal/access_tokens.json" -d 'name=example&permission=ro&scopes%5B%5D=finance&scopes%5B%5D=stats&access_token=13d435a15183e848161ec41a3f5a41b3fee6f2f5bc41ddd0308884a7c42a98cd'`
Example of response:
```json
{
 "access_token": {
  "id": 4,
  "name": "example",
  "scopes": [
   "finance",
   "stats"
  ],
  "permission": "ro",
  "value": "90127a2b42c7b59151e026e170c36abf911df05f0592bf329e63a73461d545c4"
 }
}
```
image

##### Old endpoint (deprecated)
image

### CODE REVIEW

File: app/controllers/admin/api/access_tokens_controller.rb [hunk 1]
```diff
@@ -35,8 +35,52 @@ def create
     respond_with access_token
   end
 
+  ##~ e = sapi.apis.add
+  ##~ e.path = "/admin/api/users/{user_id}/access_tokens/{id}.json"
```
Comment (hallelujah) [inline app/controllers/admin/api/access_tokens_controller.rb]: Well we should also be able I think to read an access token from its `value` not only with its `id`

WDYT @3scale/system 

It would really be impractical for an administrator to use the id instead of the value.

Comment (hallelujah) [reply app/controllers/admin/api/access_tokens_controller.rb]: Also the user_id is not necessary as it should be the same user.
We are not allowing a user to change another's token

Comment (Martouta) [reply app/controllers/admin/api/access_tokens_controller.rb]: I like this:
> Well we should also be able I think to read an access token from its value not only with its id

Not so sure about this though:
> Also the user_id is not necessary as it should be the same user.

Because it would be super weird to have the create (already in production) with one url and the show and delete with another one 🤔

Comment (Martouta) [reply app/controllers/admin/api/access_tokens_controller.rb]: Current situation: We have in production right now a `create` with a url with `user_id` that can be used with `access_token` of user and with `provider_key` of `account`. The user can only use it for itself (otherwise it is unauthorized) and the account can for any of its users.

Comment (macejmic) [reply app/controllers/admin/api/access_tokens_controller.rb]: with `provider_key` user would be able to create an access token for another user, which is not desired.. so we wether drop support for `provider_key` in this case or add `user_id` to all endpoints to mirror `create` method.. discussion in progress.

Comment (Martouta) [reply app/controllers/admin/api/access_tokens_controller.rb]: @hallelujah @guicassolato @thomasmaas @macejmic 
So what should we do about this?

Comment (Martouta) [reply app/controllers/admin/api/access_tokens_controller.rb]: The system/ruby team had a meeting about this and I will explain here what we decided. We informed product about this and I will also update the description in [THREESCALE-1766](https://issues.jboss.org/browse/THREESCALE-1766)

### What we have right now
We have only 1 API endpoint for access tokens but we have much more in the UI. The API endpoint that we have right now was created years ago before we were open source in https://github.com/3scale/system/pull/8441 but it was not documented and officially not supported and we documented it and made it public recently in https://github.com/3scale/porta/pull/430
This endpoint right now is under the URL `provider-url/admin/api/users/:user_id/access_tokens`. An officially documented that customers can authenticate through access token and we will allow to create only if the user_id belongs to the same user. Which is true but we also have (not documented and officially unsupported) the possibility to authenticate through provider_key and then we will allow to create for any user of that same account.
We also need to keep in mind that apicast may need to authenticate through provider_key 😄 

### What we we aim to have
We want it to be similar to what we have in the UI. So what we will do the next endpoints.

#### Index/Search Access Token API endpoint
Under the URL `provider-url/admin/api/access_tokens`. It will authenticate through the access_token and will return all the access tokens of the same user of the access_token if no extra parameter is sent. We will also allow the user to send a `name` string query param (like `provider-url/admin/api/access_tokens?name=example`) and we will return all the access tokens of that current user which name contains that string of the param.
For each access token we will return the following attributes.
  - scopes
  - permission
  - name
We will **not** return their values for security reasons 😄 

#### Show Access Token API endpoint
Under the URL `provider-url/admin/api/access_tokens/ID` or `provider-url/admin/api/access_tokens/VALUE`
It will authenticate through the access_token and only that same user can fetch an access token of its own. It can be fetched by `ID` or by its `value`.
We will return for the access token the same parameters as for the index (the previous endpoint).

#### Delete Access Token API endpoint
Under the URL `provider-url/admin/api/access_tokens/ID` or `provider-url/admin/api/access_tokens/VALUE`
It will authenticate through the access_token and only that same user can delete an access token of its own. It can be fetched by `ID` or by its `value`.

#### Create Access Token API endpoint
We will have a new endpoint that does exactly the same as we have, but with a different URL, under `provider-url/admin/api/access_tokens`, and this new endpoint will allow to authenticate only through access_token and not through provider_key.
The endpoint that we had until now will keep working but will be deprecated.

Comment (guicassolato) [reply app/controllers/admin/api/access_tokens_controller.rb]: An idea for the new endpoints could be namespacing them under a keyword to identify it's personal-relation stuff, such as `personal` or `new`. For example:
- `provider-url/admin/api/my/access_tokens`

Eventually, for notifications preferences:
- `provider-url/admin/api/my/notification_preferences`

I understand would means moving controllers and tests, but in case we think it's worth it, better now than having many endpoints to deprecate in the future.

Comment (Martouta) [reply app/controllers/admin/api/access_tokens_controller.rb]: Ok, I will check and let you know about it @guicassolato 👍

Comment (thomasmaas) [reply app/controllers/admin/api/access_tokens_controller.rb]: I like the idea, would probably call it `/api/personal/…` instead of `me` but if it implies too much work might not be worth it

Comment (Martouta) [reply app/controllers/admin/api/access_tokens_controller.rb]: I think like Thomas, I like it but if this has to be done relatively quickly... that may be too much work 🤔 Although maybe not, I dunno, I will try 🤔

**Comment by Martouta** [comment]
**Question for the index/search**: Can the user have so many access tokens that we should paginate?

**Comment by thomasmaas** [comment]
@Martouta theoretically that could be true but we don't paginate in ui either so lets not?

File: app/controllers/admin/api/access_tokens_controller.rb [hunk 2]
```diff
@@ -18,9 +18,10 @@ class Admin::Api::AccessTokensController < Admin::Api::BaseController
   ##~ e.responseClass = "access_token"
   #
   ##~ op = e.operations.add
+  ##~ op.deprecated = true
```
Comment (Martouta) [inline app/controllers/admin/api/access_tokens_controller.rb]: I don't see any difference in the documentation... 😞 
image

Comment (thomasmaas) [reply app/controllers/admin/api/access_tokens_controller.rb]: any difference in html? (so we can style)

Comment (thomasmaas) [reply app/controllers/admin/api/access_tokens_controller.rb]: I see that we add (deprecated) to the desc in other places.

Comment (Martouta) [reply app/controllers/admin/api/access_tokens_controller.rb]: Thanks Thomas, I will do the same then 😄 I am just waiting for more reviews first to do all the changes at once 👍

**Comment by thomasmaas** [comment]
@Martouta can we refer to the new method in the old method description?

File: app/controllers/admin/api/personal/access_tokens_controller.rb [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  ##~ op.summary   = "Access Token List"
+  ##~ op.description = "Returns the list of access tokens of the user. If the parameter name is sent, it returns only those whose name contain the string of the param"
+  ##~ op.group = "access_token"
+  #
+  ##~ op.parameters.add :name => "name", :description => "Part of the name of the access token.", :dataType => "string", :required => false, :paramType => "query"
+  #
+  ##~ op.parameters.add @parameter_access_token
+  #
+  def index
+    respond_with current_user.access_tokens.by_name(params[:name] || '')
```
Comment (hallelujah) [inline app/controllers/admin/api/personal/access_tokens_controller.rb]: I'd put `by_name(params[:name].presence)`

Comment (Martouta) [reply app/controllers/admin/api/personal/access_tokens_controller.rb]: No need imo because it checks the presence inside:
https://github.com/3scale/porta/pull/492/files/8b2364a830b5df9b6668d768c93c26b62ba9758d#diff-81b3b6c95b893b59a264c659269c7e85R12
```ruby
scope :by_name, ->(name) { name.present? ? where("name LIKE ?", "%#{name}%") : all }
```

So I could even simplify this line to:
```ruby
respond_with current_user.access_tokens.by_name(params[:name])
```

File: app/controllers/admin/api/personal/access_tokens_controller.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  end
+
+
+  private
+
+  def token
+    @token ||= begin
+      id_or_value = params[:id]
+      scoped_access_tokens = current_user.access_tokens
+      scoped_access_tokens.by_value(id_or_value) || scoped_access_tokens.find(id_or_value)
```
Comment (hallelujah) [inline app/controllers/admin/api/personal/access_tokens_controller.rb]: `by_value` looks like it will return an scope (or collection), though the implementation does not 
https://github.com/3scale/porta/blob/daf83afa159af2ce0923155aee8f6740660b17fa/app/models/access_token.rb#L119-L123
That looks strange to me. Anyway we could probably renaming it to: 

```ruby
def first_by_value_or_id(id_or_value)
  where.has { value = id_or_value | id = id_or_value }.first
end
```

File: app/controllers/admin/api/personal/access_tokens_controller.rb (hunk 3 above)
Comment (hallelujah) [reply app/controllers/admin/api/personal/access_tokens_controller.rb]: That works too

File: doc/active_docs/Account Management API.json [hunk 5]
```diff
@@ -5,9 +5,10 @@
       "responseClass": "access_token",
       "operations": [
         {
+          "deprecated": true,
           "httpMethod": "POST",
           "summary": "Access Token Create",
-          "description": "Creates an access token.",
+          "description": "Deprecated API in favor of /admin/api/personal/access_tokens . Creates an access token. Make sure to copy your new personal access token now. You will not be able to see it again as it is not stored for security reasons.",
```
Comment (hallelujah) [inline doc/active_docs/Account Management API.json]: Probably saying `[Deprecated API] Use "Personal Access Tokens Create".`

File: app/controllers/admin/api/personal/access_tokens_controller.rb (hunk 4 above)
Comment (Martouta) [reply app/controllers/admin/api/personal/access_tokens_controller.rb]: In https://github.com/3scale/porta/pull/528 😄

File: app/controllers/admin/api/personal/access_tokens_controller.rb [hunk 6]
```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+class Admin::Api::Personal::AccessTokensController < Admin::Api::Personal::BaseController
```
Comment (hallelujah) [inline app/controllers/admin/api/personal/access_tokens_controller.rb]: I wonder what is the `access_token_scopes` of this one?
`:account_management` only right?

Comment (macejmic) [reply app/controllers/admin/api/personal/access_tokens_controller.rb]: `personal` scope sounds good to me.

Comment (Martouta) [reply app/controllers/admin/api/personal/access_tokens_controller.rb]: Right now it is `:account_management`.
https://github.com/3scale/porta/blob/58ada49045c93764003ebce3a8caf863c903b306/app/controllers/admin/api/base_controller.rb#L24

What it should be, I dunno 🤔 I guess `:personal` makes sense. Let's see what the rest of the team thinks. Specially @thomasmaas

Comment (Martouta) [reply app/controllers/admin/api/personal/access_tokens_controller.rb]: Thomas also brought up (or at least from what I understood) that that the name `personal` may be confusing because all access tokens are personal 😄

Comment (macejmic) [reply app/controllers/admin/api/personal/access_tokens_controller.rb]: if there's gonna be more and more personal related endpoints i think it's a good idea to make a new scope for that.. doesn't matter on a name..

Comment (guicassolato) [reply app/controllers/admin/api/personal/access_tokens_controller.rb]: I like the idea of namespacing access tokens under "personal". Although, TBH I would prefer `my` rather than `personal`, so it would be `/admin/api/my/access_tokens.json` instead of `admin/api/personal/access_tokens.json`.

More than "account management", that's actually "self management" I think; and perhaps the same applies to Personal details and to Notification preferences – `/admin/api/my/personal_details.json`, `/admin/api/my/notification_preferences.json`.

Whenever a user performs actions over entities that only exist under his/her own account, he/she should not be required to know his/her own user ID to do it. The URLs should be more straightforward and user ID inferred from the access token used to perform the action.

That's one thing. Another thing is having a "Personal API" to group and separate all those endpoints.

I'm assuming that another admin user could also perform actions over the same resources belonging to his/her provider admin colleague, right? From the perspective of that other admin, the actions are not "self management" but indeed "account management", and they do require the user ID to be in the URL.

So, for some resources/actions, will we have two different endpoints, one for account management (with user ID in the URL) and another for self management (without user ID the URL), that do exactly the same thing. Correct? Having a "Provider API" (or "My API" or whatever) API group defined, for every pair of endpoints that do the same thing, one will be under "Provider API" while the other will remain under "Account Management API", right?

Can we apply different security policies for each one of those groups? Like, all access tokens should have RW access over whatever goes under "Provider API" I guess, while specific access tokens may have RW (or RO) access to Account Management API.

To sum up, if having the Provider API group can help us manage permissions over Account management completely independently from the permissions over Provider API, even though, in the end, some of endpoints perform exactly the same actions over the same resources, then YES. Let's create Provider API.

Comment (hallelujah) [reply app/controllers/admin/api/personal/access_tokens_controller.rb]: > Are we sure it makes sense to have a special scope for personal? Why?

I can explain:

A member which does not have `account_management` access 
image

Should still be able to create access tokens:
image

Comment (thomasmaas) [reply app/controllers/admin/api/personal/access_tokens_controller.rb]: Very true

Comment (Martouta) [reply app/controllers/admin/api/personal/access_tokens_controller.rb]: @guicassolato @hallelujah and I spoke in a call and the conclusion is:
If product agrees, let's merge this for now and we can improve it later.

File: app/controllers/admin/api/personal/access_tokens_controller.rb [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  ##~ op.group = "access_token"
+  #
+  ##~ op.parameters.add :name => "name", :description => "Name of the access token.", :dataType => "string", :required => true, :paramType => "query"
+  ##~ op.parameters.add :name => "permission", :description => "Permission of the access token. It must be either 'ro' (read only) or 'rw' (read and write).", :dataType => "string", :required => true, :paramType => "query"
+  ##~ op.parameters.add :name => "scopes", :defaultName => "scopes[]", :description => "Scope of the access token. URL-encoded array containing one or more of the possible values values. The possible values are, for a master user [\"account_management\", \"stats\"], and for a tenant user [\"finance\", \"account_management\", \"stats\"]", :dataType => "custom", :allowMultiple => true, :required => true, :paramType => "query"
+  #
+  ##~ op.parameters.add @parameter_access_token
+  #
+  def create
+    respond_with current_user.access_tokens.create(access_token_params)
```
Comment (hallelujah) [inline app/controllers/admin/api/personal/access_tokens_controller.rb]: Leaving a comment here about ability to create an access token with not allowed scope.
A user might create an access token with scope `account_management`, though he does not have `account_management` permission.

If he uses this token, he won't be authenticated because of https://github.com/3scale/porta/blob/f926ef443b037dec0af2b5d69ab03b4b2ff58e8b/lib/api_authentication/by_access_token.rb#L112

File: test/integration/api/personal/access_tokens_test.rb [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+                                                     owner: FactoryBot.create(:admin, account: @provider, admin_sections: [:partners]))
+
+      access_token_object_admin.slice(:id, :value).each_value do |id_or_value|
+        perform_request(authentication: {access_token: access_token_another_admin.value}, different_params: {id: id_or_value})
+        assert_response :not_found
+      end
+    end
+
+    test 'perform request by ID works well' do
+      return if !respond_to?(:perform_request) || !respond_to?(:assert_it_worked)
```
Comment (hallelujah) [inline test/integration/api/personal/access_tokens_test.rb]: Why is that?

File: test/integration/api/personal/access_tokens_test.rb [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+  end
+
+  class ActionsOnAnAccessToken < Admin::Api::Personal::AccessTokensTest
+    def setup
+      super
+      @access_token_object = FactoryBot.create(:access_token, owner: @admin, scopes: %w[finance])
+    end
+
+    test 'access by a non-existent ID or value responds with not_found' do
+      return unless respond_to?(:perform_request)
```
Comment (hallelujah) [inline test/integration/api/personal/access_tokens_test.rb]: I do not get why you need to return this one?

File: test/integration/api/personal/access_tokens_test.rb (hunk 8 above)
Comment (hallelujah) [reply test/integration/api/personal/access_tokens_test.rb]: OK I understand now. It looks like shared examples of rspec
You can also create the method in the sub-classes


```ruby
def perform_request
  skip 'some reason'
end
```

File: test/integration/api/personal/access_tokens_test.rb [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    authorized_member = FactoryBot.create(:member, account: @provider, admin_sections: [:partners])
+    access_token = FactoryBot.create(:access_token, owner: authorized_member, scopes: %w[account_management])
+    @access_token_object = FactoryBot.create(:access_token, owner: authorized_member)
+
+    perform_request(authentication: {access_token: access_token.value})
+    assert_it_worked
+  end
+
+  def perform_request(*)
+    skip 'Need to be implemented in its subclasses'
```
Comment (hallelujah) [inline test/integration/api/personal/access_tokens_test.rb]: `skip` or `fail` ?

Comment (Martouta) [reply test/integration/api/personal/access_tokens_test.rb]: I had no idea of the existence of this `flunk` until you mentioned. Here it says that it just fails:
https://apidock.com/ruby/MiniTest/Assertions/flunk
And that's not what I want, I want it to execute if it is implemented, which it is in all the subclasses 😄 Right now it only skips in the main class which is executed on its own as well unfortunately.

Comment (hallelujah) [reply test/integration/api/personal/access_tokens_test.rb]: What you can do then is to use modules instead of inheritance.
The only thing is you would not be able to use `test` DSL :/

File: app/controllers/admin/api/personal/access_tokens_controller.rb [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+  #
+  ##~ op = e.operations.add
+  ##~ op.httpMethod = "POST"
+  ##~ op.summary   = "Personal Access Token Create"
+  ##~ op.description = "Creates an access token. Make sure to copy your new personal access token now. You will not be able to see it again as it is not stored for security reasons."
+  ##~ op.group = "access_token"
+  #
+  ##~ op.parameters.add :name => "name", :description => "Name of the access token.", :dataType => "string", :required => true, :paramType => "query"
+  ##~ op.parameters.add :name => "permission", :description => "Permission of the access token. It must be either 'ro' (read only) or 'rw' (read and write).", :dataType => "string", :required => true, :paramType => "query"
+  ##~ op.parameters.add :name => "scopes", :defaultName => "scopes[]", :description => "Scope of the access token. URL-encoded array containing one or more of the possible values values. The possible values are, for a master user [\"account_management\", \"stats\"], and for a tenant user [\"finance\", \"account_management\", \"stats\"]", :dataType => "custom", :allowMultiple => true, :required => true, :paramType => "query"
```
[the remark this answers is not in the document]
Comment (Martouta) [reply app/controllers/admin/api/personal/access_tokens_controller.rb]: Right 👍 Too late though 😂 I will push another PR with this small change then, no worries 😄

[the remark this answers is not in the document]
Comment (Martouta) [reply app/controllers/admin/api/personal/access_tokens_controller.rb]: https://github.com/3scale/porta/pull/536 😄
